### PR TITLE
Specify InputOptions.

### PR DESCRIPTION
**Pull request by otherdaniel**
This specifies SantiizerInput, largely based on #22.

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -122,8 +127,8 @@ handle additional, application-specific use cases.
     SecureContext
   ] interface Sanitizer {
     constructor(optional SanitizerConfig config = {});
-    DOMString sanitizeToString(DOMString input);
-    DocumentFragment sanitize(DOMString input);
+    DocumentFragment sanitize(SanitizerInput input);
```
Comment (iVanlIsh) [inline index.bs:130]: SanitizerInputWithTrustedHTML?

Comment (otherdaniel) [reply index.bs:130]: I suspect not. I'm trying to handle Trusted Types integration in #49, which indeed changes this method to require TrustedHTML.

The problem I'm having is that TT is not presently an accepted standard and I think Sanitzer needs to be usable without TT. WIth 'usable' I mean not only callable from JS without TT, but also specified without relying on the TT spec draft. That is a bit awkward, honestly, but I'm not seeing any good way around this.

**Comment by mozfreddyb** [comment]
@otherdaniel I'm used to single, functional commits that can stand on their own. I feel like I'm getting mixed messages when I'm asked for review but the commits are so tiny and with inconclusive commit messages.
I'm of course totally OK with more smaller changes in tiny commits, but I'd really like their commit message to be more verbose then :-) 

I suppose we can also deal with it here by squashing them together during merge. The effective change isn't too noisy for a review, after all. Just didn't want to leave this unmentioned. :)

**Comment by otherdaniel** [comment]
Hah. I guess I'm used to different tooling here. The intent is to "squash and merge" the commit, so that the messages in the actual commit will disappear, which... is why I put zero effort into having meaningful messages there. The individual commits are effectively my edit history, and can't stand on their own.

I'll squash the branch and will re-upload.

**Review by iVanlIsh — APPROVED** [review]
LGTM

File: index.bs [hunk 2]
```diff
@@ -235,8 +252,26 @@ To <dfn>sanitize an attribute</dfn> named |attr|, run these steps:
 6. if |config| has a non-empty [=attribute allow list=] and |name| is not in |config|'s [=attribute allow list=] return 'drop'
 7. return 'keep'
 
+To <dfn>create a document fragment</dfn>
+named |fragment| from a Sanitizer |input|, run these steps:
+
+1. Switch based on |input|'s type:
+  1. if |input| is of type {{DocumentFragment}}, return |input|.
```
Comment (mozfreddyb) [inline index.bs:259]: I think we need to create a new `DocumentFragment`, in order not to copy potentially assigned observers / event listeners

File: index.bs [hunk 3]
```diff
@@ -235,8 +252,26 @@ To <dfn>sanitize an attribute</dfn> named |attr|, run these steps:
...
+  2. if |input| is of type {{Document}}
+    1. let `f` be the result of {{createDocumentFragment}}.
+    2. return the result of `input.cloneNode(input, true)`.
+  3. if |input| is of type `DOMString`, then:
```
Comment (mozfreddyb) [inline index.bs:263]: Why parse into document *and* transform into a document fragment?

**Review by mozfreddyb — COMMENTED** [review]
This is great stuff.

Originally I had the algorithm in my head along the lines below here, but I don't think that's going to work.
I had it string-based, mostly due to the fact that we can't just copy existing an DocumentFragment (observers?, event listeners?).


Gecko allows a few short cuts to pare a string directly into a document fragment, but I'm surprised there's nothing in https://dom.spec.whatwg.org/#interface-documentfragmentor even https://html.spec.whatwg.org/multipage/scripting.html#the-template-element to do that, so that's a bummer :) ...and I think that answers my question above (where I asked why you parse into document and then into fragment..)

```
1. Set |innerHTML| to the empty string
2. Switch based on |input|'s type:
  1. if |input| is of type `DOMString`, then assign |input| to |innerHTML|.
  2. if |input| is of type {{DocumentFragment}}
    1. Serialize it's node tree into a string and assign to |innerHTML|.
  3. if |input| is of type {{Document}}
    1. let |docElement| be the Document's |documentElement| property.
    2. serialize |docElement|'s into a string and assign to |innerHTML|.
3. Create a new {{DocumentFragment}}, whose host is the 
4. Invoke {{createContextualFragment}} parameters |context|, |innerHTML|,

File: index.bs (hunk 2 above)
Comment (otherdaniel) [reply index.bs]: Done.

File: index.bs (hunk 3 above)
Comment (otherdaniel) [reply index.bs:264]: As you guessed: I tried to find a parse-into-DocumentFragment algorithm in the spec, but didn't find any. Hence, the two-step process. I think our implementation can also do that directly; I just didn't find a clean way to spec it.

(Generally, I'm having a hard time finding good references for spec purposes. I could very well be overlooking something here. My understanding is that, as long as the behaviour is unambiguous we're good to go, and the implementations would certainly be free to pick simpler algorithms.)

Comment (otherdaniel) [reply index.bs:264]: As I read it, cloneNode is conveniently specified in a way to only copy the structure.

**Comment by otherdaniel** [comment]
_"I had it string-based, mostly due to the fact that we can't just copy existing an DocumentFragment (observers?, event listeners?)."_   I think [cloneNode](https://dom.spec.whatwg.org/#concept-node-clone) gives us a good option. I'm guesing string-based would also work, except for custom elements & shadow roots. `cloneNode`says _"let copy be a node that implements the same interfaces as node"_, which I'd hope handles those cases.

**Comment by mozfreddyb** [comment]
Right-o. cloneNode it is. :+1:

File: index.bs [hunk 4]
```diff
@@ -235,8 +252,26 @@ To <dfn>sanitize an attribute</dfn> named |attr|, run these steps:
...
+  1. if |input| is of type {{DocumentFragment}}, then:
+    1. let |node| refer to |input|.
+  2. if |input| is of type {{Document}}, then:
+    1. let |node| refer to |input|'s `documentElement`.
+  3. if |input| is of type `DOMString`, then:
+    1. let |node| be the result of the {{parseFromString}} algorithm
+        with |input| as first parameter (`string`),
+        and `"text/html"` as second parameter (`type`).
+2. let `f` be the result of {{createDocumentFragment}}.
+3. return the result of `f.cloneNode(node, true)`.
```
Comment (mozfreddyb) [inline index.bs:268]: I see `node . cloneNode([deep = false]) ` in spec, but you seem to be using `targetNode.cloneNode(sourceNode, true)` - What am I missing?

Comment (otherdaniel) [reply index.bs]: I think I confused myself with importNode and cloneNode. But importNode is Document-only.

 think I now formalized this with correct references.

File: index.bs (hunk 2 above)
Comment (mozfreddyb) [reply index.bs]: We're still using the input fragment here, aren't we?
Rest looks good btw

Comment (otherdaniel) [reply index.bs]: I tried to fix that in the 3rd commit, which always to |node|, and then always does a cloneNode of |node| before returning it.

(Should I squash those commits before review? I'm still a little confused about the workflow here.)

Comment (otherdaniel) [reply index.bs]: Actuallly.... :)  my version is broken. I still run cloneNode on |input|, rather than |node|. Let me fix that...

Comment (otherdaniel) [reply index.bs]: Done.

**Review by mozfreddyb — APPROVED** [review]
:ok_hand:

**Comment by otherdaniel** [comment]
Hrmpf. So... now I didn't squash the commits after all.:-((

Is it okay if I just fix that locally and push the fixed version to the repo, or does that cause issues for other users?

**Comment by mozfreddyb** [comment]
pls do it in #52, so we get a history of changes

**Pull request merged**